src/mumble/GlobalShortcut_linux.cpp

```diff
@@ -0,0 +1,187 @@
+// (like protobuf 3.7 does). As such, for now, we have to make this our last incl
+#include "Global.h"
+
+// We have to use a global 'diagnostic ignored' pragmas because
+// we still support old versions of GCC. (FreeBSD 9.3 ships with GCC 4.2)
+#if defined(__GNUC__)
+// ScreenCount(...) and so on are macros that access the private structure and
+// cast their return value using old-style-casts. Hence we suppress these warnings
+// for this section of code.
+#	pragma GCC diagnostic ignored "-Wold-style-cast"
+// XKeycodeToKeysym is deprecated.
+// For backwards compatibility reasons we want to keep using the
+// old function as long as possible. The replacement function
+// XkbKeycodeToKeysym requires the XKB extension which isn't
+// guaranteed to be present.
+#	pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#endif
+
+/**
+ * Returns a platform specific GlobalShortcutEngine object.
+ *
+ * @see GlobalShortcutX
+ * @see GlobalShortcutLinux
+ * @see GlobalShortcutMac
+ * @see GlobalShortcutWin
+ */
+GlobalShortcutEngine *GlobalShortcutEngine::platformInit() {
+	return new GlobalShortcutLinux();
+}
+
+GlobalShortcutLinux::GlobalShortcutLinux() {
+	if (g.s.bEnableEvdev) {
+		QString dir             = QLatin1String("/dev/input");
+		QFileSystemWatcher *fsw = new QFileSystemWatcher(QStringList(dir), this);
+		connect(fsw, SIGNAL(directoryChanged(const QString &)), this, SLOT(directoryChanged(const QString &)));
+		directoryChanged(dir);
+
+		if (qsKeyboards.isEmpty()) {
+			foreach (QFile *f, qmInputDevices)
+				delete f;
+			qmInputDevices.clear();
+
+			delete fsw;
+			qWarning("GlobalShortcutLinux: Unable to open any keyboard input devices under /dev/input");
+		} else {
+			return;
+		}
+	} else {
+		qWarning("GlobalShortcutLinux: No evdev support, global shortcuts disabled");
+	}
+}
+
+GlobalShortcutLinux::~GlobalShortcutLinux() {
+	wait();
+}
+
+// One of the raw /dev/input devices has ready input
+void GlobalShortcutLinux::inputReadyRead(int) {
+	if (!g.s.bEnableEvdev) {
+		return;
+	}
+
+	struct input_event ev;
+
+	if (bNeedRemap)
+		remap();
+
+	QFile *f = qobject_cast< QFile * >(sender()->parent());
+	if (!f)
+		return;
+
+	bool found = false;
+
+	while (f->read(reinterpret_cast< char * >(&ev), sizeof(ev)) == sizeof(ev)) {
+		found = true;
+		if (ev.type != EV_KEY)
+			continue;
+		bool down;
+		switch (ev.value) {
+			case 0:
+				down = false;
+				break;
+			case 1:
+				down = true;
+				break;
+			default:
+				continue;
+		}
+		int evtcode = ev.code + 8;
+		handleButton(evtcode, down);
+	}
+
+	if (!found) {
+		int fd      = f->handle();
+		int version = 0;
+		if ((ioctl(fd, EVIOCGVERSION, &version) < 0) || (((version >> 16) & 0xFF) < 1)) {
+			qWarning("GlobalShortcutLinux: Removing dead input device %s", qPrintable(f->fileName()));
+			qmInputDevices.remove(f->fileName());
+			qsKeyboards.remove(f->fileName());
+			delete f;
+		}
+	}
+}
+
+#define test_bit(bit, array) (array[bit / 8] & (1 << (bit % 8)))
```

Comment:
A short documentary comment explaining what this is supposed to do would be nice